packages/zoe/src/contracts/attestation/attestation.js

```diff
@@ -0,0 +1,158 @@
+// @ts-check
+
+import { AmountMath } from '@agoric/ertp';
+import { E } from '@agoric/eventual-send';
+import { Far } from '@agoric/marshal';
+
+import '../../../exported';
+import { setupAttestation as setupExpiringAttestation } from './expiring/expiringNFT';
+import { setupAttestation as setupReturnableAttestation } from './returnable/returnableNFT';
+import { makeStoredTime } from './storedTime';
+import { max } from './helpers';
+import { assertPrerequisites } from './prerequisites';
+import { makeGetAttMaker } from './attMaker';
+
+/**
+ * @type {ContractStartFn}
+ */
+const start = async zcf => {
+  const {
+    brands: { Underlying: underlyingBrand },
+    authority, // The authority used to confirm that the underlying assets are escrowed appropriately
+    expiringAttName,
+    returnableAttName,
+  } = zcf.getTerms();
+
+  const { assetKind: underlyingAssetKind } = await E(
+    underlyingBrand,
+  ).getDisplayInfo();
+  // AWAIT ///
+
+  const storedTime = makeStoredTime();
+  const empty = AmountMath.makeEmpty(underlyingBrand, underlyingAssetKind);
+
+  const expiringAttManagerP = setupExpiringAttestation(
+    expiringAttName, // e.g. 'BldAttGov'
+    empty,
+    zcf,
+  );
+  const returnableAttManagerP = setupReturnableAttestation(
+    returnableAttName, // e.g. 'BldAttLoc'
+    empty,
+    zcf,
+  );
+  const [expiringAttManager, returnableAttManager] = await Promise.all([
+    expiringAttManagerP,
+    returnableAttManagerP,
+  ]);
+  // AWAIT ///
+
+  /** @type {GetLienAmount} */
+  const getLienAmount = (address, currentTime) => {
+    const expiringLienAmount = expiringAttManager.getLienAmount(
+      address,
+      currentTime,
+    );
+    const returnableLienAmount = returnableAttManager.getLienAmount(address);
+    return max(expiringLienAmount, returnableLienAmount);
+  };
+
+  /** @type {GetLiened} */
+  const getLiened = (addresses, currentTime) => {
+    storedTime.updateTime(currentTime);
+    return addresses.map(address => getLienAmount(address, currentTime));
+  };
+
+  /** @type {Slashed} */
+  const slashed = (address, currentTime) => {
```

Comment:
Is this called with the currentTime, or just the time at which the slashing occurred?  If the former, should it be rejected if the currentTime is earlier than storedTime?

Reply:
i.e. I know that the storedTime will throw in that case, but should it really?

Reply:
It should be called with the latest timestamp that the caller of `slashed` knows about. So while it must be after the time slashing occurred because the caller must know about the slashing in order to make the call, it is otherwise unrelated to the slashing event itself.

I'm ok not throwing in the case that the currentTime is earlier than the storedTime.